contracts/0.8.9/WithdrawalQueueBase.sol

```diff
@@ -0,0 +1,379 @@
+        if (request.claimed) revert RequestAlreadyClaimed();
+
+
+        request.recipient = payable(_newRecipient);
+
+        requestsByRecipient[_newRecipient].add(_requestId);
+        requestsByRecipient[msg.sender].remove(_requestId);
+
+        emit WithdrawalRequestRecipientChanged(_requestId, _newRecipient, msg.sender);
+    }
+
+    /// @dev creates a new `WithdrawalRequest` in the queue
+    function _enqueue(uint256 _amountOfStETH, uint256 _amountofShares, address _recipient)
+    internal
+    returns (uint256 requestId)
+    {
+        WithdrawalRequest memory lastRequest = queue[lastRequestId];
+
+        uint256 cumulativeShares = lastRequest.cumulativeShares + _amountofShares;
+        uint256 cumulativeStETH = lastRequest.cumulativeStETH + _amountOfStETH;
+
+        lastRequestId++;
+        requestId = lastRequestId;
+        queue[requestId] = WithdrawalRequest(
+            _toUint128(cumulativeStETH),
+            _toUint128(cumulativeShares),
+            payable(_recipient),
+            _toUint64(block.number),
+            false
+        );
+        requestsByRecipient[_recipient].add(requestId);
+
+        emit WithdrawalRequested(requestId, msg.sender, _recipient, _amountOfStETH, _amountofShares);
+    }
+
+    /// @dev Finalize requests from last finalized one up to `_lastRequestIdToFinalize`
+    function _finalize(uint256 _lastRequestIdToFinalize, uint256 _amountofETH) internal {
+        if (_lastRequestIdToFinalize <= lastFinalizedRequestId) revert InvalidRequestId(_lastRequestIdToFinalize);
+        if (_lastRequestIdToFinalize > lastRequestId) revert InvalidRequestId(_lastRequestIdToFinalize);
+
+        uint128 finalizedStETH = queue[lastFinalizedRequestId].cumulativeStETH;
+        uint128 stETHToFinalize = queue[_lastRequestIdToFinalize].cumulativeStETH - finalizedStETH;
+        uint96 discountFactor = _calculateDiscountFactor(stETHToFinalize, _amountofETH);
+
+        _updateDiscountHistory(_lastRequestIdToFinalize, discountFactor);
+
+        lockedEtherAmount += _applyDiscount(stETHToFinalize, discountFactor);
+        lastFinalizedRequestId = _lastRequestIdToFinalize;
+    }
+
+    /// @dev calculates the sum of stETH and shares for all requests in (`_firstId`, `_lastId`]
+    function _batch(uint256 _firstId, uint256 _lastId)
+        internal
+        view
+        returns (uint128 amountOfStETH, uint128 amountOfShares)
+    {
+        amountOfStETH = queue[_lastId].cumulativeStETH - queue[_firstId].cumulativeStETH;
+        amountOfShares = queue[_lastId].cumulativeShares - queue[_firstId].cumulativeShares;
+    }
+
+    /// @dev returns discount factor for finalization
+    function _calculateDiscountFactor(uint256 _requestedValue, uint256 _realValue) internal pure returns (uint96) {
+        if (_requestedValue > _realValue) {
+            return _toUint96(_realValue * E27_PRECISION_BASE / _requestedValue);
```

Comment:
((_realValue * E27_PRECISION_BASE) / _requestedValue) for kogannikh meshkov

Reply:
in case when we go to this function from finalize batch we div on E27_PRECISION_BASE first and after that mul E27_PRECISION_BASE